tests/Unit/Utilities/Test_Functional.cpp

```diff
@@ -26,57 +201,202 @@ void test_get_argument() {
+                UniformCustomDistribution<DistType2>{std::get<Bound>(tup)}, x,
+                y);
+          });
+        });
+  });
 }
 
-void test_composition() {
-  CHECK(Plus<Negate<>>{}(2.0, 7.0) == -9.0);
-  CHECK(Plus<Negate<>, Identity>{}(2.0, 7.0) == 5.0);
-  CHECK(Negate<Plus<>>{}(2.0, 7.0) == -9.0);
-  CHECK(Multiplies<Plus<Plus<>, Plus<>>, Identity>{}(1, 2, 3, 4, 5) == 50);
-  CHECK(Minus<Plus<Identity, Multiplies<>>, Identity>{}(1, 2, 3, 4) == 3);
-  CHECK(Negate<Plus<Identity, Negate<>>>{}(2.0, 7.0) == 5.0);
-  CHECK(Plus<Identity, Multiplies<>>{}(1, 2, 3) == 7);
+template <typename Gen>
+void test_real_functions(Gen& gen) noexcept {
+  const Bound generic{{-50.0, 50.0}};
+  const Bound gt_one{{1.0, 100.0}};
+  const Bound positive{{.001, 100.0}};
+  const Bound small{{0.1, 5.0}};
+
+  auto generic_binaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalDivides{}, gt_one),
+                      std::make_tuple(TestFuncEvalMinus{}, generic),
+                      std::make_tuple(TestFuncEvalMultiplies{}, generic),
+                      std::make_tuple(TestFuncEvalPlus{}, generic),
+                      std::make_tuple(TestFuncEvalDivAssign{}, gt_one),
+                      std::make_tuple(TestFuncEvalMultAssign{}, generic),
+                      std::make_tuple(TestFuncEvalMinusAssign{}, generic),
+                      std::make_tuple(TestFuncEvalPlusAssign{}, generic));
+
+  auto real_binaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalAtan2{}, generic),
+                      std::make_tuple(TestFuncEvalHypot{}, generic));
+
+  auto real_unaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalCbrt{}, positive),
+                      std::make_tuple(TestFuncEvalErf{}, generic),
+                      std::make_tuple(TestFuncEvalExp10{}, small),
+                      std::make_tuple(TestFuncEvalExp2{}, small),
+                      std::make_tuple(TestFuncEvalFabs{}, generic),
+                      std::make_tuple(TestFuncEvalInvCbrt{}, gt_one),
+                      std::make_tuple(TestFuncEvalLog2{}, gt_one),
+                      std::make_tuple(TestFuncEvalStepFunction{}, generic));
+
+  tmpl::for_each<RealTypeList>([
+    &gen, &real_binaries, &generic_binaries, &real_unaries
+  ](auto x) noexcept {
+    using DistType1 =
+        typename tt::get_fundamental_type_t<typename decltype(x)::type>;
+    tuple_fold(real_unaries, [&gen, &x ](auto tup) noexcept {
+      std::get<0>(tup)(
+          gen, UniformCustomDistribution<DistType1>{std::get<Bound>(tup)}, x);
+    });
+    tmpl::for_each<RealTypeList>([&gen, &real_binaries, &generic_binaries,
+                                  &x ](auto y) noexcept {
+      using DistType2 =
+          typename tt::get_fundamental_type_t<typename decltype(y)::type>;
+      tuple_fold(std::tuple_cat(real_binaries, generic_binaries), [
+        &gen, &x, &y
+      ](auto tup) noexcept {
```

Comment:
After you mentioned the `const &` not working I realized that's because you are not doing a fold, you are doing a `for_each` with a fold. You should just do:
```
tmpl::for_each<tmpl::make_sequence<std::integral_constant<size_t, 0>, 4, tmpl::next<tmpl::_1>>>(
  [&gen, &x, &y, &tuple](auto index) noexcept {
    auto& tup = std::get<declytpe(index)::value>(tuple);
    std::get<0>(tup)(...);
  });
```